From 34a25ab9c2e07394222fa7d02e9ef6939345cd33 Mon Sep 17 00:00:00 2001
From: Ran Shalit <rshalit@habana.ai>
Date: Mon, 12 Oct 2020 11:16:01 +0300
Subject: [PATCH] [SW-14663] ppboot: handling failed verification

When validation of preboot image by ppboot fails than
all we do is just printing a message, and wfe.

Change-Id: Ief1a18fbf831db18771278d5951355c345333b4d
Signed-off-by: Ran Shalit <rshalit@habana.ai>
---

diff --git a/include/preboot.h b/include/preboot.h
index 90850d6..bac3ff0 100644
--- a/include/preboot.h
+++ b/include/preboot.h
@@ -108,6 +108,7 @@
 void hl_ts_init(void);
 bool hl_is_debug_mode(enum hl_debug_hw_func func);
 bool hl_is_jtag_enabled(void);
+void hl_enter_recovery(void);
 
 void preboot_pcie_wait_for_link(void);
 void preboot_delay(void);
diff --git a/ppboot/ppboot.c b/ppboot/ppboot.c
index 5e77e43..ac70971 100644
--- a/ppboot/ppboot.c
+++ b/ppboot/ppboot.c
@@ -205,16 +205,10 @@
 	preboot_puts("Ppboot " VERSION_STR "\n");
 }
 
-static u64 load_legacy_image(void)
+__weak void hl_enter_recovery(void)
 {
-	u64 loadaddr = PREBOOT_TEXT_BASE;
-
-	preboot_puts("No boot descriptors detected, need recovery!!!\n");
-	preboot_puts("Trying hard coded Preboot offsets (might hang)...\n");
-	ppboot_cp_preboot(loadaddr, HL_SPI_LEGACY_PREBOOT_OFF,
-			  PREBOOT_SIZE, "PREBOOT");
-
-	return loadaddr;
+	preboot_puts("No boot descriptors detected, so WFE\n");
+	wfe();
 }
 
 void ppboot_main(void)
@@ -240,8 +234,7 @@
 
 	preboot_delay();
 	ret = ppboot_load_preboot_img(&loadaddr);
-	if (!ret)
-		loadaddr = load_legacy_image();
-
-	image_start(loadaddr);
+	if (ret)
+		image_start(loadaddr);
+	hl_enter_recovery();
 }
